### ISSUE TITLE
[BUG] Matrix.mapiRows does not return a matrix

### ISSUE THREAD

**Issue by kMutagene** [fslaborg/FSharp.Stats#134]
Same is true for many of the functions starting from this line: https://github.com/fslaborg/FSharp.Stats/blob/5652e065a6c71579ac80a5278825b853b5c145c8/src/FSharp.Stats/Matrix.fs#L222

**Comment by bvenn** [comment]
Additionally, I would suggest to rename enumerateRows/Columns to mapRows/Columns and base it on Matrix.init instead of sequence expressions.

**Comment by bvenn** [comment]
@kMutagene Actually, returning a matrix leads to problems if rows/cols of the matrix should be condensed to a single value. 
The following example would fail: `Matrix.mapRows Seq.mean m`

I think the mapping function should be as generic as possible (f: Rowvector -> 'a) and not be constrained to return a sequence. It is up to the user to convert the result back into a matrix. The only thing that should be changed is the naming of enumerateRows/Cols.

**Comment by kMutagene** [comment]
Agreed. Then we should at least supply a RowVector.map function to make it easy to stay inside a Matrix<float> afterwards. Currently, i have to do this for mapping rows with an index specific function:

```fsharp
myMatrix
|> Matrix.mapiRows(fun i row -> 

	row
    |> RowVector.toArray
    |> Array.map (fun v -> v * (float i))
    |> RowVector.ofArray
)
```

while i can do this for cols:
```fsharp
tmeaRes.ConstraintPotentials
|> Matrix.mapiCols(fun i col-> 
    col
    |> Vector.map (fun v -> v * - (float i))
)
```

**Comment by HLWeil** [comment]
Hey,
just a little heads up: You could substitute 
```F#
|> matrix 
|> Matrix.transpose
```
with 
```F#
|> Matrix.ofJaggedColSeq
```
With this you can get back to the `matrix` in a (IMO) natural and straightforward way, while maintaining the generic flexibility of the matrix `map` functions as they are. 

Generally, I agree with your points though

**Comment by bvenn** [comment]
An additional comment:

When manipulating matrix rows that result in different row lengths, the `Matrix.map` operation is going to fail if the result is restricted to a matrix type. Here you have to transform it to a JaggedSeq first. Either by 

```F#
|> Matrix.transpose
|> Matrix.toJaggedArray
|> Array.map f
```

or

```F#
|> Matrix.toJaggedColArray
|> Array.map f
``` 

with the latter missing at the moment.

**Comment by bvenn** [comment]
An appropriate Matrix function still should be discussed, that enables to condense a matrix row/column by (seq<'a> -> 'b) to a single value and results in a Vector (when rowwise) and RowVector (when colwise). 
Since this operation does not require a value passing, a fold structure like prposed before seems not desirable. 

Suggestion:
`Matrix.condenseRows (fun rowvec -> f rowvec) m :Vector<'b>`

with `(f:rowvec<'a> -> 'b)`

**Comment by kMutagene** [comment]
> Since this operation does not require a value passing

What you are descripting sounds awfully like `reduce` ;) So i'd suggest `reduceRows` / `reduceCols` with the signatures as you suggested

**Comment by bvenn** [comment]
I've made some changes. What do you think about it? 🤔

**Comment by bvenn** [comment]
closed by fc39f53
